### ISSUE TITLE
Incorrectly identifies channel draining as "empty code block"

### ISSUE THREAD

**Issue by jpopadak** [mgechev/revive#386]
**Describe the bug**
`revive` incorrectly identities channel draining as an empty code block and suggests to remove it.

**To Reproduce**
1. I updated revive `go get -u github.com/mgechev/revive`
2. I run it with the following flags & configuration file:

```shell
cd $PROJECT_DIR
revive -config revive.toml .
```

```toml
ignoreGeneratedHeader = false
severity = "error"
confidence = 0.8
errorCode = 1
warningCode = 0

[rule.add-constant]
    arguments = [{maxLitCount = "3",allowStrs ="\"\"",allowInts="0,1,2"}]
[rule.argument-limit]
    arguments = [6]
[rule.atomic]
[rule.blank-imports]
[rule.bool-literal-in-expr]
[rule.call-to-gc]
[rule.constant-logical-expr]
[rule.context-as-argument]
[rule.context-keys-type]
[rule.cyclomatic]
    arguments = [10]
[rule.deep-exit]
[rule.dot-imports]
[rule.duplicated-imports]
[rule.empty-block]
[rule.error-naming]
[rule.error-return]
[rule.error-strings]
[rule.errorf]
[rule.exported]
[rule.flag-parameter]
[rule.function-result-limit]
    arguments = [4]
[rule.get-return]
[rule.if-return]
[rule.import-shadowing]
[rule.increment-decrement]
[rule.indent-error-flow]
[rule.line-length-limit]
    arguments = [120]
[rule.package-comments]
[rule.range]
[rule.range-val-in-closure]
[rule.receiver-naming]
[rule.redefines-builtin-id]
[rule.superfluous-else]
[rule.time-naming]
[rule.unexported-return]
[rule.unhandled-error]
[rule.unreachable-code]
[rule.unused-parameter]
[rule.unused-receiver]
[rule.var-declaration]
[rule.var-naming]
[rule.waitgroup-by-value]
```

A really bad example, but in our scenario, we have many routines reading from `processChan` where we send another message to have them stop processing. We reuse the channel, but we want to ensure the channel was empty from previous processing. 
```go
package main

import "sync"

const (
	chanSize = 10
	maxNums = 100
	modNum = 5
)

func main() {
	wg := sync.WaitGroup{}
	processChan := make(chan int, chanSize)

	wg.Add(1)
	go func() {
		defer wg.Done()
		for i := 0; i < maxNums; i++ {
			processChan <- i
		}
	}()

	wg.Add(1)
	go func() {
		defer wg.Done()

		for i := 0; i < maxNums; i++ {
			if !isValidNum(i) {
				break
			}
		}
	}()

	for range processChan {
	}
}

func isValidNum(i int) bool {
	return i%modNum == 0
}

```

3. Revive returns an error
>myflie.go:34:24: this block is empty, you can remove it

**Expected behavior**
No error to occur as this is channel draining.

**Logs**

**Desktop (please complete the following information):**
 - OS: macOS 10.14.6
 - Version of Go: v1.14.1

**Additional context**
We have two workarounds:
1. We can ignore the few lines across our project where we are iterating over a channel, however, it would be nicer if `revive` knew that this is channel draining. 
2. We can delete and recreate the channel, but since the channel is passed around on startup / Dependency Injection, we cannot just "create a new one and reinitialize all services" easily.

**Comment by mgechev** [comment]
Looks like we should type check to verify that we're dealing with a channel. This would slow down the rule though, but seems like the right way to go.

**Comment by chavacava** [comment]
For the sake of rule simplicity, I think these corner cases should be handled by disabling the linter in the concerned lines.
Anyway, I've pushed PR #415 with a fix based in type information.

**Comment by mgechev** [comment]
@jpopadak, because of some regressions (both stability and performance), we reverted the fix. It's still possible to recognize this pattern without type checking. This is something we'd investigate in the future.

**Comment by powerman** [comment]
Would be nice to have this fixed. I've a project with dozen of such false positives and have to disable the rule. :(

**Comment by chavacava** [comment]
branch https://github.com/mgechev/revive/tree/fix-386 implements a fix consisting in trying to identify channel draining loops by their syntax (a range without value nor key) and avoids raising a failure for such cases

branch https://github.com/mgechev/revive/tree/alleviates-386 also identifies channel draining loops by their syntax but in place of not raising a failure it raises one with a lower (0.5) confidence. 

Neither implementation uses type information

**Comment by chavacava** [comment]
Hi @feldgendler, thanks for commenting in this thread.
Rules requiring type information introduce some performance degradation thus we try to keep rules _type free_
At the time we decided to not evolve the rule because in the trade-off analysis precision vs performance degradation+complexity we considered the number of false positives not high enough to justify the use of type information to avoid them (and adding a disabling annotation in the rare cases was not too much effort)
It is something we can re-open to discussion. I'll try to rebase the branches I've pushed with the fix proposals long time ago.

**Comment by chavacava** [comment]
Yes, and yes

**Comment by feldgendler** [comment]
Would it be possible to add a fine-tuning option to the `empty-block` rule (whether to use type information) so that one could choose between accuracy and performance?

**Comment by chavacava** [comment]
I've rebased the branch #415 and now the rule does not warn on empty blocks on channel draining ranges.
@feldgendler you can try the fix and let me know if it is sufficiently precise for you.

**Comment by feldgendler** [comment]
@chavacava Yes, thank you! My code doesn't drigger `empty-block` with this fix.

**Comment by twpayne** [comment]
For info, I believe I've hit this same issue with the following code:

https://github.com/twpayne/go-heap/blob/fb16bd2df012f1689aa9c0682b0c611078c10908/heap.go#L39-L40

revive incorrectly reports that the block can be removed:

```
heap.go:39:69: empty-block: this block is empty, you can remove it (revive)
                for value, ok := h.Pop(); ok && yield(value); value, ok = h.Pop() {
                }
```

This is with revive run by golangci-lint 1.61.0, i.e. revive v1.3.9.

**Comment by chavacava** [comment]
Hi @twpayne thanks for reporting the issue.
Effectively, the rule fails to detect the side-effect of the loop predicate. Doing that kind of analysis in a complete way is beyond the capabilities of `revive` (and of any static analysis tool in the general case)
As an immediate "fix" I can just propose to deactivate the rule at that line.
I'm studying to modify the rule to just skip analyzing loops to avoid false positives.

**Comment by twpayne** [comment]
Thank you very much for the analysis and explanation. I've indeed added a `//nolint:revive` comment to suppress the warning for that particular line.

I agree that any complete static analysis is infeasible. Maybe a couple of heuristics like "does any part of the for clause, condition, or range clause contain a function call or an assignment to a variable outside the loop" might help reduce the false positives of this rule. Just a thought, I don't know how easy this would be to implement or how well it work in practice.

**Comment by feldgendler** [comment]
> Thank you very much for the analysis and explanation. I've indeed added a `//nolint:revive` comment to suppress the warning for that particular line.
> 
> I agree that any complete static analysis is infeasible. Maybe a couple of heuristics like "does any part of the for clause, condition, or range clause contain a function call or an assignment to a variable outside the loop" might help reduce the false positives of this rule. Just a thought, I don't know how easy this would be to implement or how well it work in practice.

Or `++`/`--`, or a channel op.

**Comment by chavacava** [comment]
@twpayne the rule already have a [very primitive version](https://github.com/mgechev/revive/blob/d76ffb98f6d6001e3286b2d0206df77b653829e5/rule/empty-block.go#L47) of such an heuristic but as you mention (and @feldgendler comment confirms) the cases are "infinite".
IMO the true positives this rule is spotting are very very few vs the false positives, thus I think we should just give up and do not introduce complexity to catch a negligible number of true loops with empty blocks.

**Comment by twpayne** [comment]
> IMO the true positives this rule is spotting are very very few vs the false positives, thus I think we should just give up and do not introduce complexity to catch a negligible number of true loops with empty blocks.

Ack, makes sense. Is it worth closing this issue as "not planned" or "won't fix"?

**Comment by feldgendler** [comment]
If I had a say, I would err for the side of false negatives rather than false positives. In this context, it means that an empty loop body is not reported as “empty code block” — either never reported at all, or only reported if the loop statement itself is _definitely_ side-effect-free (that is, if in doubt, don't report).

**Comment by andrew-field** [comment]
Have I misunderstood or should empty code blocks consisting of draining a channel not be flagged as an error, as of #415 ?
@chavacava 
I get an error: `prime_numbers.go:145:3: this block is empty, you can remove it`
for
```
for range primeCh {
} // Drain the primeCh channel to make sure the go routine can exit cleanly.
```
Tested with revive 1.4 and 1.7.

Full code:
```

// Copied and adapted from: tinyurl.com/gosieve
// https://youtu.be/f6kdp27TYZs

// GetPrimeNumbers returns a channel from which to siphon off the prime numbers in order, as needed.
// Send a boolean to the Done channel when finished.
// The prime sieve: Daisy-chain Filter processes.
func GetPrimeNumbers() (<-chan int, chan<- bool) {
	ch := make(chan int) // Create a new channel.
	ctx, cancel := context.WithCancel(context.Background())

	go generate(ctx, ch) // Launch Generate goroutine.

	primeCh := make(chan int) // Create return channel.
	doneCh := make(chan bool) // Create done channel.
	go func() {
		<-doneCh
		cancel()

		for range primeCh {
		} // Drain the primeCh channel to make sure the go routine can exit cleanly.
	}()

	go func() {
		defer close(primeCh)

		var prime int
		for {
			select {
			case prime = <-ch:
				primeCh <- prime
			case <-ctx.Done():
				return
			}
			ch1 := make(chan int)
			go filter(ctx, ch, ch1, prime)
			ch = ch1
		}
	}()

	return primeCh, doneCh
}

// Send the sequence 2, 3, 4, ... to channel 'ch'.
func generate(ctx context.Context, ch chan<- int) {
	for i := 2; ; i++ {
		select {
		case ch <- i: // Send 'i' to channel 'ch'.
		case <-ctx.Done():
			return
		}
	}
}

// Copy the values from channel 'in' to channel 'out',
// removing those divisible by 'prime'.
func filter(ctx context.Context, in <-chan int, out chan<- int, prime int) {
	for {
		var i int
		select {
		case i = <-in: // Receive value from 'in'.
		case <-ctx.Done():
			return
		}

		if i%prime != 0 {
			select {
			case out <- i: // Send 'i' to 'out'.
			case <-ctx.Done():
				return
			}
		}
	}
}
```

**Comment by fedorkanin** [comment]
I am experiencing the same issue as @andrew-field

**Comment by alexandear** [comment]
Documented about this limitation in #1624

**Comment by alexandear** [comment]
Another false positive example from the [samber/lo](https://github.com/samber/lo/blob/2859d7382fcda7f1ef5f525d227dfe7ae2eb9a0b/it/seq.go#L28):

```go
// Drain consumes an entire sequence.
func Drain[T any](collection iter.Seq[T]) {
	for range collection { //nolint:revive
	}
}
```

**Comment by andrew-field** [comment]
@chavacava @alexandear 
Am I right in thinking that it does now work with simple examples? i.e. revive will NOT flag draining a channel? From the docs:
```
for range processChan {
    // Intentionally empty - draining the channel
}
```

because I have some code which revive flags but, if my interpretation of the documentation is correct, it should not.

I have this benchmark function:

```
func BenchmarkGetPrimeNumbersBelowAndIncluding(b *testing.B) {
	ctx := b.Context()
	inputs := []int{10, 200, 3000, 40000, 500000}
	for _, input := range inputs {
		b.Run(fmt.Sprintf("Input: %d", input), func(b *testing.B) {
			for b.Loop() {
				for range GetPrimeNumbersBelowAndIncluding(ctx, input) { // Just iterating through the channel to benchmark the function.
				}
			}
		})
	}
}
```

where GetPrimeNumbersBelowAndIncluding has the method signature:
`func GetPrimeNumbersBelowAndIncluding(ctx context.Context, n int) <-chan int`

but I still get "prime_numbers_test.go:981:5: this block is empty, you can remove it"

This was using revive 1.15.0. This is one of five such examples.

If it makes any difference, this gives the same result:

```
ch := GetPrimeNumbersBelowAndIncluding(ctx, input)
for range ch { // Just iterating through the channel to benchmark the function.
}
```

(The code is available here: https://github.com/andrew-field/maths)

**Comment by alexandear** [comment]
@andrew-field, thank you for reporting this. I fixed it in #1710.

**Comment by andrew-field** [comment]
Ah, great! Thank you!
